CUDA: detection of CUDA on Jetson TX1 and TX2 fail

##### System information (version)
<!-- Example
- OpenCV => 4.2
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017
-->

- OpenCV => master ( 5ed6546 ) + contrib master ( 0f56e6d )
- Operating System / Platform => Jetson TX1/TX2 (Ubuntu)
- Compiler => GCC5.4.0

##### Detailed description

- cmake
```
-- General configuration for OpenCV 4.4.0-pre =====================================
--   Version control:               4.3.0-464-g5ed6546
-- 
--   Extra modules:
--     Location (extra):            /opencv_contrib/modules
--     Version control (extra):     4.3.0-69-g0f56e6d
 :
-- CUDA detected: 8.0
-- Automatic detection of CUDA generation failed. Going to build for all known architectures.
-- CUDA NVCC target flags: -ccbin;/usr/bin/cc;-gencode;arch=compute_53,code=sm_53;-gencode;arch=compute_62,code=sm_62;-gencode;arch=compute_72,code=sm_72;-D_FORCE_INLINES
```

The "Automatic detection" usually successes, but it fails, after #17526
Also, `sm72` is added, but TX1 and TX2 has CUDA 8.0, which doesn't know anything about sm72.

As a result, it fails compiling

````
nvcc fatal   : Unsupported gpu architecture 'compute_72'
CMake Error at cuda_compile_1_generated_gpu_mat.cu.o.Release.cmake:219 (message):
  Error generating
  /opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o
````

I'm not sure why the automatic detection fails, but wiping out the non-supported CC should let workaround this issue.

##### Steps to reproduce

<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   <!--
   OpenCV team works with answers.opencv.org, Stack Overflow and other communities
   to discuss problems. Tickets with question without real issue statement will be
   closed.
   -->
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   <!--
   Places to check:
   * OpenCV documentation: https://docs.opencv.org
   * FAQ page: https://github.com/opencv/opencv/wiki/FAQ
   * OpenCV forum: https://answers.opencv.org
   * OpenCV issue tracker: https://github.com/opencv/opencv/issues?q=is%3Aissue
   * Stack Overflow branch: https://stackoverflow.com/questions/tagged/opencv
   -->
 - [x] I updated to latest OpenCV version and the issue is still there
   <!--
   master branch for OpenCV 4.x and 3.4 branch for OpenCV 3.x releases.
   OpenCV team supports only latest release for each branch.
   The ticket is closed, if the problem is not reproduced with modern version.
   -->
 - [x] There is reproducer code and related data files: videos, images, onnx, etc
   <!--
   The best reproducer -- test case for OpenCV that we can add to the library.
   Recommendations for media files and binary files:
   * Try to reproduce the issue with images and videos in opencv_extra repository
     to reduce attachment size
   * Use PNG for images, if you report some CV related bug, but not image reader
     issue
   * Attach the image as archite to the ticket, if you report some reader issue.
     Image hosting services compress images and it breaks the repro code.
   * Provide ONNX file for some public model or ONNX file with with random weights,
     if you report ONNX parsing or handling issue. Architecture details diagram
     from netron tool can be very useful too. See https://lutzroeder.github.io/netron/
   -->

Along with workaround, I dug about the root cause.

- The `DETECT_ARCHS_COMMAND` [here](https://github.com/opencv/opencv/blob/5ed65460d16bc191bcf596e75d32d265405247c1/cmake/OpenCVDetectCUDA.cmake#L118)

https://github.com/opencv/opencv/blob/5ed65460d16bc191bcf596e75d32d265405247c1/cmake/OpenCVDetectCUDA.cmake#L118

```
DETECT_ARCHS_COMMAND /usr/local/cuda/bin/nvcc;-ccbin;/usr/bin/cc;/opencv-fork/cmake/checks/OpenCVDetectCudaArch.cu;--run
```

- The `DETECT_ARCHS_COMMAND` [before](https://github.com/opencv/opencv/blob/e3d502310f6ae0cad79f3474b9ea2b25705985f9/cmake/OpenCVDetectCUDA.cmake#L109) #17526

https://github.com/opencv/opencv/blob/e3d502310f6ae0cad79f3474b9ea2b25705985f9/cmake/OpenCVDetectCUDA.cmake#L109

```
DETECT_ARCHS_COMMAND /usr/local/cuda/bin/nvcc;/opencv-fork/cmake/checks/OpenCVDetectCudaArch.cu;--run
```

- When invoking both command natively on Jetson TX2 on command line
```
$ /usr/local/cuda-8.0/bin/nvcc /opencv-fork/cmake/checks/OpenCVDetectCudaArch.cu --run
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
6.2 


$ /usr/local/cuda-8.0/bin/nvcc -ccbin /usr/bin/cc /opencv-fork/cmake/checks/OpenCVDetectCudaArch.cu --run
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
/tmp/tmpxft_000018ba_00000000-16_OpenCVDetectCudaArch.o: In function `main':
tmpxft_000018ba_00000000-4_OpenCVDetectCudaArch.cudafe1.cpp:(.text+0x50): undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream(std::_Ios_Openmode)'
tmpxft_000018ba_00000000-4_OpenCVDetectCudaArch.cudafe1.cpp:(.text+0xd4): undefined reference to `std::ostream::operator<<(int)'
 :
```

So, adding `-ccbin /usr/bin/cc` to `nvcc` on Linux doesn't make sense
The actual code is based on c++, when the compiler is treating it as C.


So, in following situation, the root cause is that `CUDA_HOST_COMPILER` is added to `CUDA_NVCC_FLAGS`.
   - Jetson TX2 (TX1)
   - Ubuntu 16.04 (Aarch64)
   - CUDA 8.0
   - GCC 5.4.0
   - CMake 3.15.1

Now, going back to #17526 and interpreting the modification carefully, It's adding exactly what I mentioned above.
Adding `CUDA_HOST_COMPILER` to `CUDA_NVCC_FLAGS`.
Before I make a fix for this issue, I'd like to ask @cyyever for what is the target of PR (#17526) is targeting.
(The PR description is quite empty and I'm not sure if I should revert this PR)

@cyyever , could you comment on what's the situation you were trying to fix ?

--- a/cmake/OpenCVDetectCUDA.cmake
+++ b/cmake/OpenCVDetectCUDA.cmake
@@ -162,7 +162,12 @@ if(CUDA_FOUND)
       if(NOT _nvcc_res EQUAL 0)
         message(STATUS "Automatic detection of CUDA generation failed. Going to build for all known architectures.")
         # TX1 (5.3) TX2 (6.2) Xavier (7.2) V100 (7.0)
-        set(__cuda_arch_bin "5.3 6.2 7.2 7.0")
+        ocv_filter_available_architecture(__cuda_arch_bin
+            5.3
+            6.2
+            7.2
+            7.0
+        )
       else()
         set(__cuda_arch_bin "${_nvcc_out}")
       endif()

